365 FCS system

NikkoJT opened the pull request:
Per #365 adds the better FCS simulation module.

NikkoJT committed:
Add FCS/TC override

SamLex commented on f/fcs/fn_fcsInit.sqf:
```diff
@@ -0,0 +1,118 @@
+*/
+
+// Doesn't need to run on DS
+if (isDedicated) exitWith {};
```
This might be better as `not hasInterface`since `isDedicated` will run on a HC as well

NikkoJT replied:
Done

SamLex commented on f/fcs/fn_fcsInit.sqf:
```diff
@@ -0,0 +1,118 @@
+
+// Hit EH for FCS failure
+_vehicle addEventHandler ["HitPart",{
+		(_this select 0) params ["_target", "_shooter", "_projectile", "_position", "_velocity", "_selection", "_ammo", "_vector", "_radius", "_surfaceType", "_isDirect"];
```
If there is a direct hit (`isDirect == true`), will it always be the first element in `_this`? Otherwise, shouldn't this loop through each of the parts hit (the sub-arrays of `_this`) to check if any are direct hits?

NikkoJT replied:
I'm checking on this to be sure, but in testing the observed trigger rate seems to be appropriate for the initial impact being first, or at least a direct hit being first enough of the time to be in the right ballpark.

SamLex replied:
Fair enough. If it seems to be having the desired effect then no issue

SamLex commented on f/fcs/fn_fcsInit.sqf:
```diff
@@ -0,0 +1,118 @@
+	2. FCS failure simulation
+		When the vehicle is hit by a high-calibre weapon, there is a small chance of an FCS failure, disabling the TC override, laser rangefinder, and NVDs, and force-setting the gunner's zeroing.
+		A unit with the repair trait can reset the FCS from the gunner's seat.
+		(If and when BI adds the ability to disable turret stabilisation by script, it will do that too)
```
Is this different from `enableDirectionStabilization`?

NikkoJT replied:
Yes, `enableDirectionStabilization` only affects turrets with automatic lock-on stabilization (e.g. UAV cameras and targeting pods). At the moment there's no way to control tank-type stabilisation through sqf.

SamLex commented on f/fcs/fn_fcsFailure.sqf:
```diff
@@ -0,0 +1,44 @@
+	if !(_randomNumber == 10) exitWith { 
+		// diag_log "FCS: projectile not randomly selected" 
+	};
+	if (_target getVariable ["fcs_failure",false]) exitWith { 
```
Might make sense to move this to be the first check

NikkoJT replied:
Done

SamLex commented on f/fcs/fn_fcsFailure.sqf:
```diff
@@ -0,0 +1,44 @@
+	};
+
+	// Hit turns off the stabiliser and sets an appropriate variable on the vehicle
+	// EXCEPT IT DOESN'T TURN OFF THE STABILISER BECAUSE THAT'S NOT WHAT THE COMMAND DOES
```
:D
Out of interest, what *does* it do?

SamLex commented on f/fcs/fn_fcsPersistentEffects.sqf:
```diff
@@ -0,0 +1,52 @@
+				if !(_x == "this") then {
+					_targetMuzzle = _x;
+				};
+				_zeroCheck = _currentGunner setWeaponZeroing [_targetWeapon,_targetMuzzle,1];
```
Is `setWeaponZeroing` Global-Argument,Global-Effect?

NikkoJT replied:
The wiki helpfully does not say. I think I remember testing it and having it work as expected on remote units, but I'll have to check again to prove it.

SamLex commented on f/fcs/fn_fcsInit.sqf:
```diff
@@ -0,0 +1,118 @@
+	{}, // Code on start
+	{}, // Code on tick
+	{ 
+		params ["_target", "_caller", "_actionId", "_arguments"];
```
I think reversing the faliure effects should be extracted to a call of `f_fnc_fcsFailure` with different arguments (e.g. `["fix"] call f_fnc_fcsFailure`), or it's own function entirely

SamLex replied:
Personally like having this kind of thing "close" together in the code so if I add a new effect on the disable it obvious where the enable changes need to go. But that's really just personal preference

NikkoJT committed:
Move existing failure check to first

NikkoJT committed:
Use localWarning for all text displays

NikkoJT committed:
Change DS check to HC/DS

SamLex reviewed (dismissed):
Looks good to me.

NikkoJT committed:
Unlock FCS after 4 seconds

Commander's override now lasts 4 seconds at maximum to prevent automated pixelhunting from locking out the gunner.

NikkoJT committed:
Add light mode

The FCS system can now be flagged to only add the driver's brake override action and skip the rest.

NikkoJT commented:
Made a couple of small improvements since we're still waiting for other stuff for release.

NikkoJT committed:
Fix missing setVariable

NikkoJT committed:
Add briefing tab generator

NikkoJT committed:
Clarify briefing

NikkoJT committed:
F3-ify variable names

SamLex commented on f/fcs/fn_fcsBriefing.sqf:
```diff
@@ -0,0 +1,34 @@
+
+if (!hasInterface) exitWith {}; // Exit if not a player.
+
+_fcs = player createDiaryRecord ["diary", ["F3 Enhanced FCS","
```
This is nice! My only concern is that this will appear at the _top_ of the Briefing tab, whereas it makes more sense at/near the bottom (e.g. with credits and administration). Could `f_fnc_fcsInit` be moved to before the briefing in `init.sqf`.

An alternative would be to add this to it's own subject with `createDiarySubject`.

NikkoJT replied:
This is consistent with other "supplemental" items such as Mission Conditions, F3 Nametags, and the mapClickTeleport HALO instructions. I could move it, but I'm not all that bothered about it to be honest.

NikkoJT replied:
In local testing the FCS briefing tab seems to appear at the bottom anyway.

SamLex replied:
Hmm, so it does. Probably `spawn` for briefing vs `call` for this (and maybe also the `waitUntil` in briefing. Guess if it works it works

SamLex commented on f/fcs/fn_fcsBriefing.sqf:
```diff
@@ -0,0 +1,34 @@
+"]];
+
+// Set a variable so this won't be generated again by subsequent inits
+f3_fcs_briefingDone = true;
```
Minor: I think the convention elsewhere is `f_var_` rather than `f3_`. But either is probably fine

SamLex commented on f/fcs/fn_fcsFailure.sqf:
```diff
@@ -9,6 +9,10 @@ params ["_target","_projectile","_isDirect"];
 	private _warheadType = "";
 	private _randomNumber = 0;
+	
+	if (_target getVariable ["fcs_failure",false]) exitWith { 
```
```suggestion
	if (_target getVariable ["f3_fcs_failure",false]) exitWith { 
```
If you want to keep `f3_`

SamLex replied:
Or
```suggestion
	if (_target getVariable ["f_var_fcs_failure",false]) exitWith { 
```

SamLex commented on f/fcs/fn_fcsInit.sqf:
```diff
@@ -24,83 +24,93 @@
-	""	
-];
+// If the FCS briefing tab hasn't been added already, add it.
+if (isNil "f3_fcs_briefingDone") then {
```
Minor: my personal preference would be for this to be an `exitWith` at the top of `f/fcs/fn_fcsBriefing.sqf`, instead of it being the responsibility of the caller to check.

NikkoJT replied:
I wanted to do it this way because fcsInit can be run _a lot_ depending on the vehicles in the mission, so saving the function call is [probably] better for performance.

SamLex commented on f/fcs/fn_fcsInit.sqf:
```diff
@@ -24,83 +24,93 @@
+};
 
+// Only do this part in the default mode. If in the alternate mode, skip this and only add the driver action.
+if (_mode) then {
```
At this point, would the driver actions not be better as a separate component? Adding the `_mode` makes this feel a little overloaded (and would remove the need for the `Note:` in the briefing tab)

SamLex commented on f/fcs/fn_fcsInit.sqf:
```diff
@@ -24,83 +24,93 @@
+if (_vehicle getVariable ["f3_fcs_hasDriverAction",false]) exitWith { diag_log "FCS: tried to add driver action on something that already has it"};
 
 // Not FCS but I'm putting it in here
 _vehicle addAction [
```
Question: continuation of the idea of this being a separate component, is there a reason not to add this to every (land) vehicle? With that the action could be added to the player instead and just act on any vehicle they are the driver of.

NikkoJT replied:
It is possible, although unarmed vehicles don't really benefit from turning off the autobrakes, which is why I didn't bother with it. While it's not "FCS" in the technical sense, it's part of a suite of improvements targeted at armed vehicles in particular - for example, the advanced FCS stuff is useless for technicals but the brake override is useful. So this got added from the side of "expand the gunnery improvements" rather than as a separate thing on its own.

So I guess it comes down to whether to consider it a utility for armed vehicles or whether it's actually a separate thing all vehicles should have.

NikkoJT committed:
Update f/fcs/fn_fcsFailure.sqf

Co-authored-by: Euan Hunter <SamLex@users.noreply.github.com>

NikkoJT committed:
Rename vars to standard

NikkoJT committed:
Move brakes to separate component

NikkoJT committed:
Add F3 thermals handling

NikkoJT committed:
Fix wrong command target

NikkoJT committed:
Fix brake action condition

SamLex commented on f/fcs/fn_fcsInit.sqf:
```diff
@@ -0,0 +1,110 @@
+		[_target,_projectile,_isDirect] call f_fnc_fcsFailure;
+	}];
+	
+_vehicle setVariable ["fcs_hasEH",true];
```
```suggestion
_vehicle setVariable ["f_var_fcs_hasEH",true];
```

SamLex commented on f/fcs/fn_fcsInit.sqf:
```diff
@@ -0,0 +1,110 @@
+// Doesn't need to run on DS or HC
+if !(hasInterface) exitWith {};
+
+params ["_vehicle",["_mode",true]];
```
```suggestion
params ["_vehicle"];
```
`_mode` isn't needed now the brake override is it's own thing (thank you!)

SamLex commented on f/fcs/fn_fcsLocalWarning.sqf:
```diff
@@ -0,0 +1,18 @@
+
+params ["_text","_delay"];
+
+if (cameraView == "GUNNER") then {
```
This _probably_ should call [disableSerialization](https://community.bistudio.com/wiki/disableSerialization) before calling the UI commands, if I'm reading the wiki right.

NikkoJT replied:
Normally there's a script error if this is going to be a problem, and this has never caused one. Adding disableSerialization anyway though.

SamLex reviewed (dismissed):
Looks good to me! Two small typos suggestion fixes, and a possible best practise thing. Looking forward to having this in 3.5.7 and, by proxy, more missions

NikkoJT committed:
fix variable name

Co-authored-by: Euan Hunter <SamLex@users.noreply.github.com>

NikkoJT committed:
Remove mode switch vestiges

Co-authored-by: Euan Hunter <SamLex@users.noreply.github.com>

NikkoJT committed:
Disable serialization for UI things

NikkoJT committed:
fix system not counting Tandem HEAT warheads

NikkoJT committed:
Merge branch 'dev' into FCS-system

NikkoJT commented:
Was conflicted with the radio branch. That's resolved, just needs reapproval
